imbi/endpoints/project_ssm_configuration.py

```diff
@@ -0,0 +1,211 @@
+
+        imbi_user = await self.get_current_user()
+        env = os.environ.get('ENVIRONMENT', 'production').lower()
+        google_integration = await oauth2.OAuth2Integration.by_name(
+            self.application, f'google-{env}')
+        tokens = await google_integration.get_user_tokens(imbi_user)
+        if len(tokens) == 0:
+            raise web.HTTPError(
+                401,
+                reason=
+                'Not authorized to access SSM in this namespace & environment')
+        google_tokens = tokens[0]
+
+        ssm_path_prefix = self.path_prefix(
+            self.application.settings['project_configuration']
+            ['ssm_prefix_template'], project_info['project_slug'],
+            project_info['project_type_slug'], project_info['namespace_slug'])
+
+        role_arn_exists = False
+        params_by_path = {}
+        for environment in project_info['environments']:
+            result = await self.postgres_execute(
+                self.GET_ROLE_ARN_SQL, {
+                    'environment': environment,
+                    'namespace_id': project_info['namespace_id']
+                }, 'get-role-arn')
+            if not result.row:
+                continue
+            else:
+                role_arn_exists = True
+            role_arn = result.row['role_arn']
+
+            success, data = await self.get_aws_credentials(
+                role_arn, imbi_user, google_tokens.id_token,
+                google_tokens.refresh_token)
+            if not success:
+                if data == 'AccessDenied':
+                    raise web.HTTPError(
+                        401,
+                        reason=
+                        'Not authorized to access SSM in this namespace & environment'
+                    )
+                else:
+                    self.logger.error('Unexpected AWS credential failure: %s',
+                                      data)
+                    raise web.HTTPError(
+                        500,
+                        reason='Unexpected failure fetching AWS credentials')
+            params = aws.get_parameters_by_path(ssm_path_prefix,
+                                                data['access_key_id'],
+                                                data['secret_access_key'],
+                                                data['session_token'])
+            for param in params:
+                if param['Name'] in params_by_path:
+                    params_by_path[
+                        param['Name']]['values'][environment] = param['Value']
+                else:
+                    params_by_path[param['Name']] = {
+                        'type': param['Type'],
+                        'values': {
+                            environment: param['Value']
+                        }
+                    }
+        if not role_arn_exists:
+            raise web.HTTPError(
+                401,
```

Comment:
I'm on the fence about this one... I might return `null` in this case though `null` versus `[]` can be a bit confusing. Semantically, I like `null` since an empty array is not falsy in JS .... but it is falsy in other languages 😞 

The problem is that not having an ARN configured isn't something that the client can change (eg, 4xx vs 5xx) so if you are going to return a failure, then a 500 is likely appropriate... but a retry isn't going to solve this either. One could make a case for a `400 Bad Request`?

Reply:
After a bit of reading, I agree it should be a 500-level response, since the client can't do anything about it. While not necessarily its typical usage, I think 501 actually makes the most sense here.

Reply:
The more that I think about this, returning an empty array response is probably the most appropriate thing. The endpoint is retrieving the configured parameters. If there is at least one ARN configured but no data is found, we return an empty array. Not having an environment ARN configured might as well return the same thing ... _in both cases, no SSM parameters exist for the project_.

That change will also simplify the calling code since they will not have to worry about having an error returned for the same end result -- no configuration data available. In the cases where we have no data because of an AWS error, then we will return a failure. The result is that the user code doesn't have to be concerned with what is likely a misconfiguration since the outcome is identical. We should tick a metric and/or log something here.

Reply:
Updated to return `[]` and log that it couldn't find any role arn for any of the environments for the project